### PR TITLE
Create podvm images via github actions

### PR DESCRIPTION
**Pull request by bpradipt**

### CODE REVIEW

**Comment by bpradipt** [comment]
@jtumber-ibm ptal.
I'm yet to run the "Create Pod VM Image" action due to quota issues under my github account.

**Comment by jtumber-ibm** [comment]
Got it running in my account. Is the plan to make quay.io/confidential-containers/cloud-api-adaptor-podvm-* image public and potential users have to do something like

```
docker create --name temp-container quay.io/confidential-containers/cloud-api-adaptor-podvm-ubuntu:latest
docker cp temp-container:/podvm.qcow2 podvm.qcow2
```

or should there be another step to add that file as a release asset?

**Comment by bpradipt** [comment]
We can do both. Have the image in quay.io. And also another action to create release artefacts..

**Comment by jtumber-ibm** [comment]
> @jtumber-ibm one thing missing from this PR is creating podvm images for different providers. I have to create a matrix runner for all the providers. Is there anything else missing ?

Is there anything that makes the images specfic for any provider? During testing I was able to take the qcow image that was created for libvirt (using the Docker build) and import that into ibmcloud and use that as is.

**Comment by bpradipt** [comment]
> > @jtumber-ibm one thing missing from this PR is creating podvm images for different providers. I have to create a matrix runner for all the providers. Is there anything else missing ?
> 
> Is there anything that makes the images specfic for any provider? During testing I was able to take the qcow image that was created for libvirt (using the Docker build) and import that into ibmcloud and use that as is.

The agent-protocol-forwarder is not really provider specific. But I don't remember if there is anything else that's provider specific. I'll need to check...

**Comment by jtumber-ibm** [comment]
I was able to run the github action (with edits to run against a registry I have write access too). And it successfully built the image.

**Comment by bpradipt** [comment]
Successful job run details 
- https://github.com/bpradipt/cloud-api-adaptor/actions/runs/3876170752
- https://github.com/bpradipt/cloud-api-adaptor/actions/runs/3876225116

Few notable changes
1. Shortened the image names as podvm-builder-{distro} and podvm_{provider}-{distro}
2. Matrix runner to create images per provider. This is to ensure flexibility and optimisations that could be provider specific. For example using CentOS ec2 qcow2 images when creating aws image. Although not a must, but it gives us the flexibility.

**Comment by bpradipt** [comment]
Example set of instructions to extract the Qcow2

```
PODVM_IMG=quay.io/confidential-containers/podvm-aws-centos:latest
docker run $PODVM_IMG bash 2>/dev/null
CID=$(docker ps -n 1 -q)
QCOW2=$(docker export $CID | tar t | grep  podvm)
docker cp $CID:/$QCOW2 .
```

**Comment by bpradipt** [comment]
@jtumber-ibm I couldn't add you to the review list.  Kindly review
cc @stevenhorsman

**Comment by jtumber-ibm** [comment]
> 

Pulling the ibmcloud image, will setup a cluster to check the resulting image works successfully

**Comment by jtumber-ibm** [comment]
Quick test of uploading produce qcow2 image to release

```yaml
    - name: Extract Asset
      run: |
        CID=$(docker create --name temp-container quay.io/confidential-containers/podvm-${{ matrix.provider }}-${{ matrix.os }}:latest /bin/sh)
        podvm=$(docker export $CID | tar t | grep podvm)
        docker cp temp-container:/$podvm podvm-${{ matrix.provider }}-${{ matrix.os }}.qcow2
        docker rm temp-container
    - name: Upload Asset
      uses: softprops/action-gh-release@v1
      with:
        files: podvm-${{ matrix.provider }}-${{ matrix.os }}.qcow2
```

Created https://github.com/jtumber-ibm/cloud-api-adaptor/releases/download/upload-release-0.2/podvm-ibmcloud-ubuntu.qcow2

File: .github/workflows/podvm.yaml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      with:
+        registry: quay.io
+        username: ${{ secrets.QUAY_USERNAME }}
+        password: ${{ secrets.QUAY_PASSWORD }}
+
+    - name: Build and push
+      uses: docker/build-push-action@v3
+      with:
+        tags: |
+          quay.io/confidential-containers/podvm-${{ matrix.provider }}-${{ matrix.os }}:latest
```
Comment (jtumber-ibm) [inline .github/workflows/podvm.yaml:55]: If adding the provider to matrix, should add to image name?

Comment (bpradipt) [reply .github/workflows/podvm.yaml:55]: Are you referring to the qcow2 image name inside the podvm container image ?  If yes then we'll need to change the Makefile.  Probably as a separate PR .

**Comment by bpradipt** [comment]
aws image is not working :-( 

> Quick test of uploading produce qcow2 image to release
> 
> ```yaml
>     - name: Extract Asset
>       run: |
>         CID=$(docker create --name temp-container quay.io/confidential-containers/podvm-${{ matrix.provider }}-${{ matrix.os }}:latest /bin/sh)
>         podvm=$(docker export $CID | tar t | grep podvm)
>         docker cp temp-container:/$podvm podvm-${{ matrix.provider }}-${{ matrix.os }}.qcow2
>         docker rm temp-container
>     - name: Upload Asset
>       uses: softprops/action-gh-release@v1
>       with:
>         files: podvm-${{ matrix.provider }}-${{ matrix.os }}.qcow2
> ```
> 
> Created https://github.com/jtumber-ibm/cloud-api-adaptor/releases/download/upload-release-0.2/podvm-ibmcloud-ubuntu.qcow2

cool.. Do you want me to add this in the same workflow or you want to create a new PR and add it as a separate workflow ?

File: .github/workflows/podvm.yaml (hunk 1 above)
Comment (jtumber-ibm) [reply .github/workflows/podvm.yaml:55]: Yeah sorry that wasn't clear, I mixed up two of the problems I had. I think this is fine as is.

**Comment by jtumber-ibm** [comment]
> aws image is not working :-(
> 
> cool.. Do you want me to add this in the same workflow or you want to create a new PR and add it as a separate workflow ?

Could the image not be working due to #505 ?

Do you think that renaming the image like `podvm-${{ matrix.provider }}-${{ matrix.os }}.qcow2` is good enough or should be add in the git hash like the original name has? So that the files stay unique.

If you're happy to could you add the same workflow, might need updates to make the `temp-container` unique for each provider/os combo.

**Comment by bpradipt** [comment]
> > aws image is not working :-(
> > cool.. Do you want me to add this in the same workflow or you want to create a new PR and add it as a separate workflow ?
> 
> Could the image not be working due to #505 ?
> 

No.. In my case it's using the wrong base centos qcow2 for AWS. I did mention this in my comments on provider specific image, but forgot to set it in the action :-( 

> Do you think that renaming the image like `podvm-${{ matrix.provider }}-${{ matrix.os }}.qcow2` is good enough or should be add in the git hash like the original name has? So that the files stay unique.
> 

Hmm. good point. May be better to add the git hash to the image to keep it unique. 

> If you're happy to could you add the same workflow, might need updates to make the `temp-container` unique for each provider/os combo.

Sure, let me add it.

**Comment by bpradipt** [comment]
@jtumber-ibm @snir911 incorporated the comments and also added a step to create release artifacts. Ptal

**Review by jtumber-ibm — APPROVED** [review]
lgtm, thanks @bpradipt

File: .github/workflows/podvm_builder.yaml [hunk 2]
```diff
@@ -0,0 +1,59 @@
+name: Create Pod VM Builder Image
+on:
+  release:
+    types: [created]
```
Comment (snir911) [inline .github/workflows/podvm_builder.yaml:4]: maybe worth adding manual option? not critical ATM though

File: .github/workflows/podvm.yaml [hunk 3]
```diff
@@ -61,3 +61,17 @@ jobs:
...
         build-args: |
           "CLOUD_PROVIDER=${{ matrix.provider }}"
+
+    - name: Extract Asset
+      run: |
+        UUID=$(cat /proc/sys/kernel/random/uuid | cut -c 1-8)
+        CONTAINER_NAME=temp-container-${UUID}
+        CID=$(docker create --name $CONTAINER_NAME quay.io/confidential-containers/podvm-${{ matrix.provider }}-${{ matrix.os }}:${{ github.sha }} /bin/sh)
+        podvm=$(docker export $CID | tar t | grep podvm)
+        docker cp $CONTAINER_NAME:/$podvm podvm-${{ matrix.provider }}-${{ matrix.os }}-${{ github.sha }}.qcow2
```
Comment (snir911) [inline .github/workflows/podvm.yaml:71]: nit: `podman save ${IMG_ID} | tar -xO --no-wildcards-match-slash '*.tar' | tar -x` worked for me but i'm fine with that as well

File: podvm/Dockerfile.podvm [hunk 4]
```diff
@@ -4,17 +4,22 @@
...
-ARG BUILDER_IMG="quay.io/confidential-containers/cloud-api-adaptor-podvm-builder"
+ARG BUILDER_IMG="quay.io/confidential-containers/podvm-builder-ubuntu"
 
 FROM ${BUILDER_IMG} AS podvm_builder
 
 ARG CLOUD_PROVIDER=libvirt
 ARG PODVM_DISTRO=ubuntu
 
-# If these are not provided as build args then the
-# default values assigned in the Makefile will get used
```
Comment (snir911) [inline podvm/Dockerfile.podvm:15]: should these be removed from the makefile?

**Review by snir911 — APPROVED** [review]
added some comments but LGTM overall, thanks!

File: podvm/Dockerfile.podvm (hunk 4 above)
Comment (bpradipt) [reply podvm/Dockerfile.podvm:15]: For running non-container builds these might be needed.

File: .github/workflows/podvm_builder.yaml (hunk 2 above)
Comment (bpradipt) [reply .github/workflows/podvm_builder.yaml:4]: We can add manual running of workflow in subsequent PRs

**Pull request merged**